File: branches/feature/Android_Update_Image/src/org/openremote/android/console/view/ButtonView.java
/* OpenRemote, the Home of the Digital Home.
* Copyright 2008-2010, OpenRemote Inc.
*
* See the contributors.txt file in the distribution for a
* full listing of individual contributors.
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU Affero General Public License as
* published by the Free Software Foundation, either version 3 of the
* License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU Affero General Public License for more details.
*
* You should have received a copy of the GNU Affero General Public License
* along with this program. If not, see <http://www.gnu.org/licenses/>.
*/
package org.openremote.android.console.view;

import java.util.Timer;
import java.util.TimerTask;

import org.openremote.android.console.Constants;
import org.openremote.android.console.bindings.ORButton;
import org.openremote.android.console.model.ListenerConstant;
import org.openremote.android.console.model.OREvent;
import org.openremote.android.console.model.OREventListener;
import org.openremote.android.console.model.ORListenerManager;
import org.openremote.android.console.util.ImageUtil;

import android.content.Context;
import android.graphics.drawable.BitmapDrawable;
import android.os.Handler;
import android.os.Message;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Button;
import android.widget.FrameLayout;

public class ButtonView extends ControlView {

   private Button uiButton;
   private BitmapDrawable defaultImage;
   private BitmapDrawable pressedImage;
   
   /** The Constant REPEAT_CMD_INTERVAL. */
   public final static long REPEAT_CMD_INTERVAL = 300;
   
   public ButtonView(Context context, ORButton button) {
      super(context);
      setComponent(button);
      if (button != null) {
         uiButton  = new Button(context);
         initButton(button);
      }
   }
   
   /**
    * Inits the button.
    * Add listener to make the button send command and navigate to.
    * If pressed, send command; if up, navigate to. It also can 
    * repeat send command if pressed and not up.  
    * 
    * @param button the button
    */
   private void initButton(final ORButton button) {
      final int width = button.getFrameWidth();
      final int height = button.getFrameHeight();
      uiButton.setId(button.getComponentId());
      uiButton.setText(button.getName());
      uiButton.setTextSize(Constants.DEFAULT_FONT_SIZE);
      uiButton.setLayoutParams(new FrameLayout.LayoutParams(width, height));
      if (button.getDefaultImage() != null) {
         final String defaultImageName = button.getDefaultImage().getSrc();
         defaultImage = ImageUtil.createClipedDrawableFromPath(Constants.FILE_FOLDER_PATH + defaultImageName, width, height);
         ORListenerManager.getInstance().addOREventListener(ListenerConstant.LISTENER_IMAGE_CHANGE_FORMAT + defaultImageName, new OREventListener() {
            public void handleEvent(OREvent event) {
               defaultImage = null;
               defaultImage = ImageUtil.createClipedDrawableFromPath(Constants.FILE_FOLDER_PATH + defaultImageName, width, height);
               handler.sendEmptyMessage(0);
            }
         });
         if (defaultImage != null) {
            uiButton.setBackgroundDrawable(defaultImage);
         }
      }
      if (button.getPressedImage() != null) {
         final String pressedImageName = button.getPressedImage().getSrc();
         pressedImage = ImageUtil.createClipedDrawableFromPath(Constants.FILE_FOLDER_PATH + pressedImageName, width, height);
         ORListenerManager.getInstance().addOREventListener(ListenerConstant.LISTENER_IMAGE_CHANGE_FORMAT + pressedImageName, new OREventListener() {
            public void handleEvent(OREvent event) {
               pressedImage = null;
               pressedImage = ImageUtil.createClipedDrawableFromPath(Constants.FILE_FOLDER_PATH + pressedImageName, width, height);
            }
         });
      }
      View.OnTouchListener touchListener = new OnTouchListener() {
         public boolean onTouch(View v, MotionEvent event) {
            if (event.getAction() == MotionEvent.ACTION_DOWN) {
               cancelTimer();
               if (pressedImage != null) {
                  uiButton.setBackgroundDrawable(pressedImage);
               } else if (defaultImage != null) {
                  defaultImage.setAlpha(200);
                  uiButton.setBackgroundDrawable(defaultImage);
               }
               if (button.isHasControlCommand()) {
                  sendCommand();
                  if (button.isRepeat()) {
                     Timer timer = new Timer();
                     timer.schedule(new TimerTask() {
                        public void run() {
                           sendCommand();
                        }
                     }, REPEAT_CMD_INTERVAL, REPEAT_CMD_INTERVAL);
                     setTimer(timer);
                  }
               }
            } else if (event.getAction() == MotionEvent.ACTION_UP) {
               cancelTimer();
               if (defaultImage != null) {
                  defaultImage.setAlpha(255);
                  uiButton.setBackgroundDrawable(defaultImage);
               }
               if (button.getNavigate() != null) {
                  uiButton.setPressed(false);
                  ORListenerManager.getInstance().notifyOREventListener(ListenerConstant.ListenerNavigateTo, button.getNavigate());
               }
            }
            return false;
         }
      };
      uiButton.setOnTouchListener(touchListener);
      
      addView(uiButton);
   }

   private void sendCommand() {
      sendCommandRequest("click");
   }
   
   /** The handler is for updating default image. */
   private Handler handler = new Handler() {
      @Override
      public void handleMessage(Message msg) {
         if (msg.what == 0) {
            uiButton.setBackgroundDrawable(defaultImage);
         }
         super.handleMessage(msg);
      }
  };
}

File: workspace/ebariaux/iPad_Console_2_0_0_Betas_ButtonEnhancement/Classes/ButtonView.m
/* OpenRemote, the Home of the Digital Home.
 *  * Copyright 2008-2011, OpenRemote Inc-2009, OpenRemote Inc.
 * 
 * See the contributors.txt file in the distribution for a
 * full listing of individual contributors.
 * 
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3.0 of
 * the License, or (at your option) any later version.
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

#import "ButtonView.h"
#import "Control.h"
#import "DirectoryDefinition.h"
#import "ServerDefinition.h"
#import "ViewHelper.h"
#import "NotificationConstant.h"
#import "ClippedUIImage.h"

@interface ButtonView ()

- (void)createButton;
- (void)controlButtonUp:(id)sender;
- (void)controlButtonDown:(id)sender;
- (void)sendPressCommand:(id)sender;

@end

@implementation ButtonView

@synthesize uiButton, uiImage, uiImagePressed;

#pragma mark Private methods

//Create button according to control and add tap event
- (void)createButton {
	if (uiButton != nil) {
		[uiButton release];
		uiButton = nil;
	}
	uiButton = [[UIButton buttonWithType:UIButtonTypeCustom] retain];
	
	[uiButton addTarget:self action:@selector(controlButtonDown:) forControlEvents:UIControlEventTouchDown];	
	[uiButton addTarget:self action:@selector(controlButtonUp:) forControlEvents:UIControlEventTouchUpOutside];	
	[uiButton addTarget:self action:@selector(controlButtonUp:) forControlEvents:UIControlEventTouchUpInside];

	[self addSubview:uiButton];
}

// Event handler for button up.
- (void) controlButtonUp:(id)sender {
	[self cancelTimer];
	Button *button = (Button *)component;
    
    
    
    if (button.hasShortReleaseCommand) {
    }

    
    
	if (button.navigate) {
		[[NSNotificationCenter defaultCenter] postNotificationName:NotificationNavigateTo object:button.navigate];
	}
}

// Event handler for button down.
- (void) controlButtonDown:(id)sender {
	[self cancelTimer];
	
	Button *button = (Button *)component;
	if (button.hasPressCommand == YES) {
		[self sendPressCommand:nil];
	 	if (button.repeat == YES ) {			
			controlTimer = [NSTimer scheduledTimerWithTimeInterval:button.repeatDelay / 1000.0	target:self selector:@selector(sendCommand:) userInfo:nil repeats:YES];			
		}
	}
    if (button.hasLongPressCommand || button.hasLongReleaseCommand) {
        // Set-up timer to detect when this becomes a long press
    }

}

// Send control command to remote controller server.
- (void) sendPressCommand:(id)sender {
	[self	sendCommandRequest:@"click"];
}

- (void)cancelTimer {
	if (controlTimer) {
		[controlTimer invalidate];
	}
	controlTimer = nil;
}

#pragma mark Override the methods of superclass(ComponentView)

- (void)initView {	
	[self createButton];
	
	Button *button = (Button *)component;
	if (button.defaultImage) {
		uiImage = [[UIImage alloc] initWithContentsOfFile:[[DirectoryDefinition imageCacheFolder] stringByAppendingPathComponent:button.defaultImage.src]];
		uiImagePressed = [[UIImage alloc] initWithContentsOfFile:[[DirectoryDefinition imageCacheFolder] stringByAppendingPathComponent:button.pressedImage.src]];	
		ClippedUIImage *clippedUIImage = [[ClippedUIImage alloc] initWithUIImage:uiImage dependingOnUIView:self imageAlignToView:IMAGE_ABSOLUTE_ALIGN_TO_VIEW];		
		[uiButton setBackgroundImage:clippedUIImage forState:UIControlStateNormal];
        [clippedUIImage release];
		if (uiImagePressed) {
			ClippedUIImage *clippedUIImagePressed = [[ClippedUIImage alloc] initWithUIImage:uiImagePressed dependingOnUIView:self imageAlignToView:IMAGE_ABSOLUTE_ALIGN_TO_VIEW];
			[uiButton setBackgroundImage:clippedUIImagePressed forState:UIControlStateHighlighted];
            [clippedUIImagePressed release];
		}
		//use top-left alignment
		[uiButton setFrame:CGRectMake(0, 0, clippedUIImage.size.width, clippedUIImage.size.height)];
	} else {
		[uiButton setFrame:[self bounds]];
		UIImage *buttonImage = [[UIImage imageNamed:@"button.png"] stretchableImageWithLeftCapWidth:20 topCapHeight:29];
		[uiButton setBackgroundImage:buttonImage forState:UIControlStateNormal];
	}
	
	uiButton.titleLabel.font = [UIFont boldSystemFontOfSize:13];
	uiButton.titleLabel.lineBreakMode = UILineBreakModeTailTruncation;
	//[uiButton setTitleShadowColor:[UIColor grayColor] forState:UIControlStateNormal];
	//uiButton.titleLabel.shadowOffset = CGSizeMake(0, -2);
	[uiButton setTitle:button.name forState:UIControlStateNormal];	
}

#pragma mark dealloc

- (void)dealloc {
	[uiImage  release];
	[uiImagePressed release];
	[uiButton release];
	
    [controlTimer release];

    [super dealloc];
}

#pragma mark ORControllerCommandSenderDelegate implementation

- (void)commandSendFailed
{
    [super commandSendFailed];
    [self cancelTimer];
}


@end
